keps/sig-apps/0008-sidecarcontainer.md

```diff
@@ -0,0 +1,116 @@
+
+### Goals
+
+Solve issue [25908](https://github.com/kubernetes/kubernetes/issues/25908) so that sidecars can be easily handled without needing modifications to the container itself
+
+### Non-Goals
+
+Changing the way jobs are handled
+
+## Proposal
+
+Create a way to define containers as sidecars, this will be an additional field to the Container Spec: `sidecar: true`.
+e.g:
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: myapp-pod
+  labels:
+    app: myapp
+spec:
+  containers:
+  - name: myapp
+    image: myapp
+    command: ['do something']
+  - name: sidecar
+    image: sidecar-image
+    sidecar: true
+    command: ["do something to help my app"]
+
+```
+These sidecar containers will be handled the same way as a non-sidecar container in almost all respects apart from the fact that when all non-sidecar containers have permanently terminated all running sidecar containers will also be terminated with a reason of `Completed`. The sidecars will be terminated by receiving a SIGTERM and will obey the `terminationGracePeriodSeconds` defined on the pod before being sent a SIGKILL. Sidecar containers will be terminated in the same way as non-sidecars when a pod is deleted.
+
+For example, in a job you can have as many sidecars as you like and all the running sidecars will be terminated when all the non-sidecar containers have terminated, causing your job to finish.
+If all the non-sidecars `Completed` the pod status will be `Succeeded`, if any of the non-sidecars terminated with `Error` the pod status will be `Failed` and if any of the sidecars terminated with `Error` before the non-sidecars terminated, the pod status will be `Failed`.
+
+### Implementation Details/Notes/Constraints [optional]
+
+//TODO
+
+### Risks and Mitigations
+
+You could set all containers to be `sidecar: true`, this seems wrong, so maybe the api should do a validation check that at least one container is not a sidecar.
+Init containers would be able to have `sidecar: true` applied to them as it's an additional field to the container spec, there's no situation where this would make sense as init containers are executed sequentially. We could get around this by having the api throw a validation error if you try to use this field on an init container.
+
+## Graduation Criteria
+
+//TODO
+
+## Implementation History
+
+//TODO
+
+Major milestones in the life cycle of a KEP should be tracked in `Implementation History`.
+Major milestones might include
+
+- the `Summary` and `Motivation` sections being merged signaling SIG acceptance
+- the `Proposal` section being merged signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+
+
+## Alternatives
```

Comment:
How about instead of distinguishing only between normal containers and sidecar containers, make it more like a dependency graph. That would also make it possible to have an application container, which depend on sidecar container 1, which in turn depends on sidecar container 2? (can't think of an example though 🙃 )

Spec could look like:

```
containers:
  - name: db-proxy
    # ...
  - name: my-app
    dependencies:
      - name: db-proxy
    # ...
```

This will make Kubernetes start `db-proxy` before `my-app` starts, and kill `db-proxy` after `my-app` stops (no matter how it is stopped.

(I'm relatively new to Kubernetes, so I don't know if this has already been discussed elsewhere)

Reply:
This sounds very difficult to reason about -- the Pod container set isn't natively hierarchial, so it's not obvious how recursive inter-container dependencies would work in terms of lifetimes.

Implementation of a `sidecar: true` field doesn't preclude implementing dependencies at a later date, so I think it's worth deferring this idea to a later time.

Reply:
I'd prefer to leave the complexity of managing a full graph to something like https://github.com/knative/build-pipeline